software/controller/lib/debug/commands.h

```diff
@@ -0,0 +1,219 @@
+// We don't actually have a boot mode yet, but its only a matter of time.
+// Once we start doing things like updating firmware (not through a debugger),
+// we will need a separate boot loader image to ensure graceful recovery.
+class ModeHandler : public Handler {
+public:
+  ModeHandler() = default;
+  ErrorCode Process(Context *context) override {
+    context->response_length = 1;
+    context->response[0] = 0;
+    return ErrorCode::kNone;
+  }
+};
+
+// Console command.
+// This allows the contents of the print buffer data to be read.
+// The print buffer is where strings go when you call debug.Print
+class ConsoleHandler : public Handler {
+public:
+  ConsoleHandler() = default;
+
+  ErrorCode Process(Context *context) override {
+    // No data needs to be passed in to this command.
+
+    // Read bytes from the print buffer until I hit my
+    // max or the buffer is empty.
+
+    uint32_t count;
+    for (count = 0; count < context->max_response_length; count++) {
+      std::optional<uint8_t> next_char = debug.GetNextCharFromBuffer();
+      if (next_char == std::nullopt)
+        break;
+      context->response[count] = *next_char;
+    }
+
+    context->response_length = count;
+    return ErrorCode::kNone;
+  }
+};
+
+// Peek command.
+// Allows us to read raw memory values.
+//
+// The data passed to the command consists of a 32-bit starting address and
+// a 16-bit byte count.
+class PeekHandler : public Handler {
+public:
+  PeekHandler() = default;
+  ErrorCode Process(Context *context) override {
+    // The data passed to this command consists of a 4 byte address
+    // and a two byte count of how many bytes to return
+    // Length should be exactly 6, but if more data is passed
+    // I'll just ignore it.
+    if (context->request_length < 6)
+      return ErrorCode::kMissingData;
+
+    uint32_t address = u8_to_u32(&context->request[0]);
+    uint32_t count = u8_to_u16(&context->request[4]);
+
+    // Limit the number of output bytes based on buffer size
+    if (count > context->max_response_length)
+      count = context->max_response_length;
+
+    // Some registers can't handle byte accesses, so rather then just
+    // use a simple memcpy here I will do 32-bit or 16-bit accesses
+    // as long as both the address and count are aligned to 32 or 16 bits.
+    if (!(address & 3) && !(count & 3)) {
+      uint32_t *ptr = reinterpret_cast<uint32_t *>(address);
+
+      for (int byte_number = 0; byte_number < count / 4; byte_number++) {
+        uint32_t value = *ptr++;
+        u32_to_u8(value, &context->response[4 * byte_number]);
+      }
+    } else if (!(address & 1) && !(count & 1)) {
+      uint16_t *ptr = reinterpret_cast<uint16_t *>(address);
+      for (int byte_number = 0; byte_number < count / 2; byte_number++) {
+        uint16_t value = *ptr++;
+        u16_to_u8(value, &context->response[2 * byte_number]);
+      }
+    } else {
+      memcpy(context->response, reinterpret_cast<void *>(address), count);
```

Comment:
Is the expectation that `memcpy` copies bytes at a time? That's typically not true.

Reply:
I'm inclined to think we should remove the `Peek` and `Poke` handlers.  In most cases, `memcpy` is safe. Where its not, I think the developer should be explicit in how they are accessing the hardware.
What do you think @asmodai27 ?

Reply:
Just to be clear, I didn't add this code, merely moved it from another file.
Having those peek and poke commands seems important to me, whatever the machinery behind them.
I really don't know if this attempt at sanitizing `memcpy` changes anything to the behaviour. All I know is right now it does work this way, and to be honest I am not 100% sure how to cover all cases when testing the alternative on the hardware.

Reply:
It doesn't appear to be called from anywhere, but I guess if it moved from elsewhere, let's just leave it alone for now.

Reply:
Like all debug interface command handlers, it is called when the firmware receives the corresponding command on the serial link, which in this case is a user input in the `controller_debug` script. I have used poke a couple times when working on I2C interface, and I use peek on a regular basis.